wp/lib/bap_wp/src/constraint.ml

```diff
@@ -209,18 +209,43 @@ let rec eval_aux ?stats:(stats = init_stats) (constr : t) (olds : z3_expr list)
       (eval_aux ~stats:stats c2 olds news ctx)
   | Clause (hyps, concs) ->
     let eval_conjunction conj =
-      List.map conj ~f:(fun c -> eval_aux ~stats:stats c olds news ctx)
-      |> Bool.mk_and ctx
+      if List.length conj = 1 then 
+        eval_aux ~stats:stats (List.hd_exn conj) olds news ctx (* This is tail recursive. Avoids And node. *)
+      else
+        List.map conj ~f:(fun c -> eval_aux ~stats:stats c olds news ctx)
+        |> Bool.mk_and ctx
     in
-    let concs_expr = eval_conjunction concs in
     if List.is_empty hyps then
-      concs_expr
+      eval_conjunction concs(* possibly tail recursive? *)
     else
+      let concs_expr = eval_conjunction concs in
       let hyps_expr = eval_conjunction hyps in
       Bool.mk_implies ctx hyps_expr concs_expr
   | Subst (c, o, n) ->
     let n' = List.map n ~f:(fun x -> Expr.substitute x olds news) in
-    eval_aux ~stats:stats c (olds @ o) (news @ n') ctx
```

Comment:
Perhaps a silly comment, but can we keep the above code as a comment explaining that that is the notional semantics of the following block of code?

Reply:
That makes sense